src/main/java/uk/ac/ebi/ampt2d/accession/service/MonotonicAccessionGenerator.java

```diff
@@ -0,0 +1,241 @@
+        long[] accessions = new long[totalAccessionsToGenerate];
+        increaseAvailableRangesUntilSizeIs(totalAccessionsToGenerate);
+
+        int i = 0;
+        while (i != totalAccessionsToGenerate) {
+            int remainingAccessionsToGenerate = totalAccessionsToGenerate - i;
+            MonotonicRange monotonicRange = pollNextMonotonicRange(remainingAccessionsToGenerate);
+            int idsInRange = monotonicRange.getTotalOfValues();
+            System.arraycopy(monotonicRange.getIds(), 0, accessions, i, idsInRange);
+            i += idsInRange;
+        }
+
+        generatedIds.addAll(LongStream.of(accessions).boxed().collect(Collectors.toList()));
+        return accessions;
+    }
+
+    public synchronized List<MonotonicRange> generateAccessionRanges(int totalAccessionsToGenerate) {
+        List<MonotonicRange> accessionRanges = new ArrayList<>();
+        increaseAvailableRangesUntilSizeIs(totalAccessionsToGenerate);
+
+        int i = 0;
+        while (i != totalAccessionsToGenerate) {
+            int remainingAccessionsToGenerate = totalAccessionsToGenerate - i;
+            MonotonicRange monotonicRange = pollNextMonotonicRange(remainingAccessionsToGenerate);
+            accessionRanges.add(monotonicRange);
+            i += monotonicRange.getTotalOfValues();
+        }
+        return accessionRanges;
+    }
+
+    /**
+     * Polls the next monotonic range to use, if the next available range contains more values than needed, splits
+     * the range and returns the excess to available ranges.
+     *
+     * @param totalValuesNeeded
+     * @return
+     */
+    private synchronized MonotonicRange pollNextMonotonicRange(int totalValuesNeeded) {
+        MonotonicRange monotonicRange = availableRanges.poll();
+        if (monotonicRange.getTotalOfValues() > totalValuesNeeded) {
+            Pair<MonotonicRange, MonotonicRange> splitResult = monotonicRange.split(totalValuesNeeded);
+            monotonicRange = splitResult.getFirst();
+            availableRanges.add(splitResult.getSecond());
+        }
+        return monotonicRange;
+    }
+
+    /**
+     * Ensures that the available ranges queue hold @param totalAccessionsToGenerate or more elements
+     *
+     * @param totalAccessionsToGenerate
+     */
+    private synchronized void increaseAvailableRangesUntilSizeIs(int totalAccessionsToGenerate) {
+        while (availableRanges.getTotalOfValuesInQueue() < totalAccessionsToGenerate) {
+            try {
+                ExponentialBackOff.execute(() -> reserveNewBlock(categoryId, instanceId, blockSize), 10, 30);
+            } catch (RuntimeException e) {
+                // Ignore, max backoff have been reached, we will try again until we can reserve blocks
+            }
+        }
+    }
+
+    /**
+     * Initialize the block service loading previously reserved blocks
+     */
+    @Override
+    public void afterPropertiesSet() {
```

Comment:
It could be possible to make this a setter and call it at the end of the constructor.